### PR TITLE
DI Support ctor injection & added NLogResolveException

### PR DESCRIPTION
**Pull request by 304NotModified**
- DI Support ctor injection 
-  added NLogResolveException

### CODE REVIEW

File: src/NLog/Config/ServiceRepository.cs [hunk 1]
```diff
@@ -79,32 +82,97 @@ public void RegisterType(Type type, ConfigurationItemCreator objectResolver)
...
+        public object ResolveService(Type itemType)
         {
             var createInstance = CreateInstance;
             if (createInstance != null)
             {
                 return createInstance(itemType);
             }
 
-            return DefaultResolveInstance(itemType);
+            return DefaultResolveInstance(itemType, new HashSet<Type>());
```
Comment (snakefoot) [inline src/NLog/Config/ServiceRepository.cs]: Please don't allocate upfront.

File: src/NLog/Config/ServiceRepository.cs [hunk 2]
```diff
@@ -62,7 +65,7 @@ internal ServiceRepository(bool resetGlobalCache = false)
                 ConfigurationItemFactory.Default = null;    //build new global factory
 
             this.RegisterDefaults();
-            CreateInstance = DefaultResolveInstance;
+            CreateInstance = itemType => DefaultResolveInstance(itemType, new HashSet<Type>());
```
Comment (snakefoot) [inline src/NLog/Config/ServiceRepository.cs]: Please don't allocate upfront.

File: src/NLog/Config/ServiceRepository.cs [hunk 3]
```diff
@@ -79,32 +82,97 @@ public void RegisterType(Type type, ConfigurationItemCreator objectResolver)
...
             {
+                InternalLogger.Trace("Resolve {0} done", itemType.FullName);
                 return objectResolver(itemType);
             }
 
             try
             {
-                return Activator.CreateInstance(itemType);
+                //todo DI cache/find upfront
+                var defaultConstructor = itemType.GetConstructor(Type.EmptyTypes);
```
Comment (snakefoot) [inline src/NLog/Config/ServiceRepository.cs]: Can this lookup be cached?

File: src/NLog/Config/ServiceRepository.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: fixed

File: src/NLog/Config/ServiceRepository.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: fixed

(PS: I will always try to fix it everywhere)

File: src/NLog/Config/ServiceRepository.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: Yes, it's a TODO ;)

Maybe another PR?

File: tests/NLog.UnitTests/Config/ServiceRepositoryTests.cs [hunk 4]
```diff
@@ -44,51 +46,81 @@ public class ServiceRepositoryTests : NLogTestBase
         [Fact]
         public void SideBySideLogFactoryExternalInterfaceTest()
         {
-            // Stage
-            LogFactory logFactory1 = new LogFactory();
-            logFactory1.ServiceRepository.RegisterType(typeof(IMyPrettyInterface), (t) => new MyPrettyImplementation() { Test = nameof(logFactory1) });
+            // Arrange
+            var logFactory1 = new LogFactory();
+            const string name1 = "name1";
+            logFactory1.ServiceRepository.RegisterType(typeof(IMyPrettyInterface), (t) => new MyPrettyImplementation() { Test = name1 });
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Config/ServiceRepositoryTests.cs]: @snakefoot PS; the `nameof` confused me a lot here. So create a separate variable "name")

File: src/NLog/Config/ServiceRepository.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Config/ServiceRepository.cs]: Sorry cannot do dependency injection without proper performance :)

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: This isn't to the master /dev yet.

;)

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: we could add it after this. I could do it, or maybe you if you like.

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: trying to keeping the PRS small

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: PS with the default ctor is as fast as before ;)

Comment (snakefoot) [reply src/NLog/Config/ServiceRepository.cs]: No this is your baby :)

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: For performance we need much more, e.g. another Register so we could do the reflection stuff on registering. :)

Comment (snakefoot) [reply src/NLog/Config/ServiceRepository.cs]: Not talking about the performance of calling. It was the `itemType.GetConstructor`-lookup. `Activator.CreateInstance` performs caching. You need to re-implement that here instead.

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: OK, so I should prefer Activator.CreateInstance here? 

> Not talking about the performance of calling actual constructor. It was the itemType.GetConstructor-lookup

I'm also not. You could do the lookup on register.

Comment (snakefoot) [reply src/NLog/Config/ServiceRepository.cs]: No when you have performed the constructor-lookup, then it doesn't make sense to ask `Activator.CreateInstance` to also perform constructor-lookup.

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: OK, will do in another PR

PS: We need at the end a final PR to dev, then we can soon/refactor also.

Comment (snakefoot) [reply src/NLog/Config/ServiceRepository.cs]: I'm still not sure why we have this feature-branch. Not like dev is released tomorrow. But I will wait until you are done. Then I will look into the LogEventInfo-part.

Comment (304NotModified) [reply src/NLog/Config/ServiceRepository.cs]: Because it pollutes dev and I try to keep PRs small

**Pull request merged**